packages/transaction-messages/src/fee-payer.ts

```diff
@@ -11,10 +11,10 @@ export function setTransactionMessageFeePayer<
     TTransaction extends BaseTransactionMessage,
 >(
     feePayer: Address<TFeePayerAddress>,
-    transaction: TTransaction | (ITransactionMessageWithFeePayer<string> & TTransaction),
-): ITransactionMessageWithFeePayer<TFeePayerAddress> & TTransaction {
+    transaction: TTransaction,
+): ITransactionMessageWithFeePayer<TFeePayerAddress> & Omit<TTransaction, 'feePayer'> {
     if ('feePayer' in transaction && feePayer === transaction.feePayer) {
-        return transaction as ITransactionMessageWithFeePayer<TFeePayerAddress> & TTransaction;
+        return transaction as ITransactionMessageWithFeePayer<TFeePayerAddress> & Omit<TTransaction, 'feePayer'>;
     }
     const out = {
         ...transaction,
```

Comment:
@lorisleiva, similarly, if the fee payer address changes, this code won't blank out any pre-existing `feePayerSigner` should one exist. Can this cause problems if I go from:

* Loris(A)/Phantom, to
* Loris(B)/NO_SIGNER

The Phantom `feePayerSigner` will still be on the message, right?

Reply:
Oooh this is a bit more tricky because this part of the code is totally unaware of the signer API which sits on top of the transaction layer.

I can see this being a source of confusion though if someone tried to update or "clear" a fee payer signer using a fee payer address. Especially as the signer API will focus on the signer property and not the address property so it'll look like that last update was never called.

I think it's probably beneficial here to clear the fee payer signer attribute in a way that doesn't tight this package to the signers package. Like "I don't know what that `feePayerSigner` thing is, but if you've got that in your object, we'll take it out".

That way, we fix this potential confusion whilst keeping the packages untangled. Wdyt?

Reply:
We sort of deserve this for creating two mutually exclusive property combinations instead of a union type.

We could fix this once and for all by making `feePayer` a discriminated union.

```
export interface ITransactionMessageWithFeePayer<TAddress extends string = string> {
    readonly feePayer: Readonly<{__type: 'address', address: Address<TAddress>}>;
}
export interface ITransactionMessageWithFeePayerSigner<
    TAddress extends string = string,
    TSigner extends TransactionSigner<TAddress> = TransactionSigner<TAddress>,
> {
    readonly feePayer: Readonly<{__type: 'signer'}> & TSigner;
}
```

That would be a whole bunch of surgery, but we'd end up with something that always had `address` on it as we have now, no matter if it was a signer or not.

Reply:
Filed as #2899.